src/librustc_lint/lib.rs

```diff
@@ -165,7 +166,8 @@ pub fn register_builtins(store: &mut lint::LintStore, sess: Option<&Session>) {
                     "nonstandard_style",
                     NON_CAMEL_CASE_TYPES,
                     NON_SNAKE_CASE,
-                    NON_UPPER_CASE_GLOBALS);
+                    NON_UPPER_CASE_GLOBALS,
+                    MISLEADING_CONSTANT_PATTERNS);
```

Comment:
cc @Manishearth 

1. what do you think of the name? It follows our rule of needing to be readable as "allow misleading constant patterns"
2. Should this maybe just live in clippy, since it's not really a bug, but more a stylistic confusing problem

Reply:
In thinking about (2) a bit more, I do wonder if this lint really carries its weight. The rationale for the lint in #7526 is to avoid misleading patterns where constants look like bindings. However, to even get in this situation you'd have to explicitly `#[allow(non_upper_case_globals)]` on your const or static in the first place. For cases where you have to use a non-upper case global, then this lint really just serves as an extra hurdle that you're going to allow anyways.

I think it might suffice to include a blurb about the misleading patterns in the rationale for `non_upper_case_globals`: "patterns containing consts or statics that are not upper cased look like bindings, and can cause difficult-to-spot bugs". Maybe as a note?

Reply:
The issue is not declaring and using such a constant within one crate, but exporting such constants in a crate and then pattern matching on the name e.g. by accident, when you actually wanted a binding.

Reply:
Ah, that makes sense. I'll wait for @Manishearth to weigh in before making any more changes.